chore: use testcontainers-go instead of containertest

mdelapenya opened the pull request:
**Description:** <Describe what has changed.>
<!--Ex. Fixing a bug - Describe the bug and how this fixes the issue.
Ex. Adding a feature - Explain what this achieves.-->

This PR is replacing the internal package for managing containers lifecycle in tests with testcontainers-go. Following the same procedure as in #16209, where we updated the Redis receiver tests, we are updating:

- zookeeper receiver
- memcached receiver
- docker stats receiver
- aerospike receiver
- kafka metrics receiver
- docker observer extension

**Link to tracking Issue:**  #9939

**Testing:** <Describe what testing was performed and which tests were added.>

I ran each module tests with:

```shell
cd $the_module
go test -timeout 600s -v -tags integration
```

**Documentation:** No docs were updated

**Follow-ups**
After merging this PR, I think the `containertest` package can be removed/deprecated.

mdelapenya committed:
chore: use testcontainers-go in zookeper receiver

mdelapenya committed:
fix: remove forgotten dependency in redis receiver

mdelapenya committed:
chore: use testcontainers-go in memcached receiver

mdelapenya committed:
chore: use testcontainers-go in docker stats receiver

mdelapenya committed:
chore: use testcontainers-go in aerospike receiver

mdelapenya committed:
chore: use testcontainers-go in kafka metrics receiver

mdelapenya committed:
chore: use testcontainers-go in docker observer extension

mdelapenya commented on receiver/redisreceiver/go.mod:
```diff
@@ -67,7 +67,5 @@ require (
 
 replace github.com/open-telemetry/opentelemetry-collector-contrib/internal/common => ../../internal/common
 
-replace github.com/open-telemetry/opentelemetry-collector-contrib/internal/containertest => ../../internal/containertest
```
I forgot to remove this line in #16209

mdelapenya commented:
It seems the same error than yesterday happened at the windows pipeline. Is there anything I can do to troubleshoot it?

djaglowski commented:
> It seems the same error than yesterday happened at the windows pipeline. Is there anything I can do to troubleshoot it?

I opened https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/16267 to track the issue. I don't recall seeing this one in particular before, but we do deal with [flaky tests](https://github.com/open-telemetry/opentelemetry-collector-contrib/labels/flaky%20test) much more frequently than we would like.

Ideally we would resolve every issue but these are often difficult to reproduce/debug. We try to at least document any failures in a corresponding issue and note reoccurrences as well. If nothing else, this provides more information about what may be happening. If a failure occurs frequently we will disable the test until it is resolved. We could certainly use more help with these flaky tests. Any help or ideas for managing are welcome.

djaglowski reviewed (approved):
LGTM. Thanks again @mdelapenya.

mdelapenya commented:
The error seems a real one in `dockerstatsreceiver.TestRemovedContainerRemovesRecordsIntegration`.

Currently looking at it 👀

mdelapenya committed:
fix: consider the current Docker state in tests

mdelapenya commented:
@djaglowski I did my best trying to understand the test suite 🤞 Result is https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/16264/commits/f01b5bda345a765b27828afedabf7726dc289b61

I noticed the test failed because the assertions was expecting an isolated Docker execution, and it could be the case that other containers are running.

I'd recommend checking the current assertions regarding Docker events/stats, as they could be influenced by the existence of, i.e., the Ryuk container or other pending-to-be-pruned containers. Please take a look at https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/16264/commits/f01b5bda345a765b27828afedabf7726dc289b61#diff-ca81407f807d8cfe041b4a8ad51b991f51e0fcc20bfd8ff6bad32b9b9ac9e0ddR224 and https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/16264/commits/f01b5bda345a765b27828afedabf7726dc289b61#diff-ca81407f807d8cfe041b4a8ad51b991f51e0fcc20bfd8ff6bad32b9b9ac9e0ddR247, where I'm checking the initial count of running containers, and expecting just one more in the assertion.

Same issue in https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/16264/commits/f01b5bda345a765b27828afedabf7726dc289b61#diff-ca81407f807d8cfe041b4a8ad51b991f51e0fcc20bfd8ff6bad32b9b9ac9e0ddR255, where I created a stub Container from the container ID.

Hopefully I guessed correctly 😅

antonblock reviewed (approved):
Aerospike changes look good and pass integration tests

mdelapenya commented:
Mongodb receiver tests failing:
```
--- FAIL: TestMongodbIntegration (0.00s)
    --- FAIL: TestMongodbIntegration/Running_mongodb_4.0 (16.87s)
    --- PASS: TestMongodbIntegration/Running_mongodb_4.2 (79.81s)
    --- PASS: TestMongodbIntegration/Running_mongodb_4.4_as_LPU (81.02s)
    --- PASS: TestMongodbIntegration/Running_mongodb_5.0 (76.70s)
FAIL
coverage: 73.3% of statements
FAIL	github.com/open-telemetry/opentelemetry-collector-contrib/receiver/mongodbreceiver	156.561s
?   	github.com/open-telemetry/opentelemetry-collector-contrib/receiver/mongodbreceiver/internal/metadata	[no test files]
FAIL
make[1]: *** [../../Makefile.Common:68: do-integration-tests-with-cover] Error 1
make: *** [Makefile:188: receiver/mongodbreceiver] Error 2
```

We could probably be facing "integrated tests", or tests that could eventually depend on each other's state. Will take a look at it once I have time.

mdelapenya committed:
fix: decouple mongoDB integration tests

The container requests where using same fixed, exposed port, instead of
using the random one. In this commit we are calculating the exposed port
in the getContainer method, returning the endpoint for the container
in order to be used by each test

mdelapenya commented:
I noticed the mongoDB test examples were using a fixed port to be exposed, and that port was the same for mongo [4.0](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/16264/commits/ffe8ca5aecb578c513645114b3ba6caa66a008af#diff-3967959f6b8f81572ebce39791ba27a28b2d7e27850d1990861cd219c3fbf736L46) and [4.2](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/16264/commits/ffe8ca5aecb578c513645114b3ba6caa66a008af#diff-3967959f6b8f81572ebce39791ba27a28b2d7e27850d1990861cd219c3fbf736L54).

I updated the function to return the mongo container to retrieve both the container and its endpoint, calculating the dynamic, random port internally.

Hope this fixes the build 🤞

mdelapenya commented:
The build failed because of the fpm install: https://github.com/open-telemetry/opentelemetry-collector-contrib/actions/runs/3454696043/jobs/5766407878

```
Run gem install --no-document fpm -v 1.11.0
  gem install --no-document fpm -v 1.11.0
  shell: /usr/bin/bash -e {0}
  env:
    TEST_RESULTS: testbed/tests/results/junit/results.xml
    SEGMENT_DOWNLOAD_TIMEOUT_MINS: 5
ERROR:  While executing gem ... (Gem::RemoteFetcher::FetchError)
    too many connection resets (https://rubygems.org/quick/Marshal.[4](https://github.com/open-telemetry/opentelemetry-collector-contrib/actions/runs/3454696043/jobs/5766407878#step:4:4).[8](https://github.com/open-telemetry/opentelemetry-collector-contrib/actions/runs/3454696043/jobs/5766407878#step:4:9)/fpm-1.11.0.gemspec.rz)
Error: Process completed with exit code 1.
```

Probably a retry on the step could help here: https://github.com/orgs/community/discussions/27121#discussioncomment-3254700

djaglowski commented:
@mdelapenya, really appreciate your efforts on this. 

Given that the mongo changes go beyond what is necessary for the others, would you be willing to open a separate PR for those?

mdelapenya commented:
> Given that the mongo changes go beyond what is necessary for the others, would you be willing to open a separate PR for that component?

Sure, I can revert the commit for MongoDB and send it in a separate PR. I'm worried if, without the commit, the PR does not get merged because the mongo tests could eventually fail for the inconsistency of the used ports. If you agree with that, then't it's fine for me too :)

mdelapenya committed:
Revert "fix: decouple mongoDB integration tests"

This reverts commit ffe8ca5aecb578c513645114b3ba6caa66a008af.

mdelapenya committed:
Merge branch 'main' into receivers-testcontainers-go
* main:
  [chore] dependabot updates Tue Nov 15 00:17:56 UTC 2022 (#16300)
  [receiver/rabbitmq] Fix flaky integration test (#16240)
  [Docker Stats Receiver] Add @jamesmoessis as code owner (#16253)
  [receiver/mongodbatlas] Alerts poll mode check hostname and port (#16286)
  [chore] update jaeger dep (#16287)
  [pkg/stanza] Fix issue where specifying a non-existent storage extension caused panic during shutdown. (#16213)
  [pkg/ottl] Rename ottllogs to ottllog (#16242)
  Register featuregate, otherwise does not work (#16269)
  [receiver/elasticsearch]: add fielddata memory size metrics on index level (#14922)
  [chore] remove what looks to be debugging code from instanaexporter (#16258)
